examples/server/server.cpp

```diff
@@ -2561,8 +2561,35 @@ int main(int argc, char **argv)
         server_state current_state = state.load();
         switch(current_state) {
             case SERVER_STATE_READY:
-                res.set_content(R"({"status": "ok"})", "application/json");
-                res.status = 200; // HTTP OK
+                if (llama.all_slots_are_idle) {
+                    res.set_content(R"({"status": "ok"})", "application/json");
+                    res.status = 200; // HTTP OK
+                } else {
+                    int available_slots = 0;
+                    int processing_slots = 0;
+                    for (llama_client_slot & slot : llama.slots) {
+                        if (slot.available()) {
+                            available_slots++;
+                        } else {
+                            processing_slots++;
+                        }
+                    }
+                    if (available_slots > 0) {
+                        json health = {
+                                {"status",           "ok"},
+                                {"slots_idle",       available_slots},
+                                {"slots_processing", processing_slots}};
+                        res.set_content(health.dump(), "application/json");
+                        res.status = 200; // HTTP OK
+                    } else {
+                        json health = {
+                                {"status",           "no slot available"},
+                                {"slots_idle",       available_slots},
+                                {"slots_processing", processing_slots}};
+                        res.set_content(health.dump(), "application/json");
+                        res.status = 503; // HTTP Service Unavailable
```

Comment:
@phymbert thanks for introducing this additional metadata to the health check!

one nit: it seems unidiomatic for health to return an error status code for an expected and error-free state. in practice, for a local inference server with a single slot (the default behavior), this is particularly unintuitive.

while the server _is_ busy wrt inference, it can happily process health check requests — why return an error (`5xx`) status code, rather than a success (request understood and processed just fine) along with the actual information desired, the count of available slots (`0`)?

503 or 409 conflict make more sense to me for `/completion` or chat completion requests — their request can genuinely not be processed. but the health check returning 5xx codes during normal operation feels wrong to me. the server is not unhealthy by any metric.

it seems this is [not an uncommon point of bike shedding](https://stackoverflow.com/a/40424959) so I will happily work around this behavior if i'm in the minority, but wanted to share in case there was any other agreement to this effect.

happy to put up a patch if so!

Reply:
hi @brittlewis12, thanks for your feedback.

My primary goal is to point a [kubernetes readiness probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-readiness-probes) to the health endpoint. This way, the server will not receive new incoming request but they will be routed to another available pod. It does not mean the server is down, but as `503` says: it is overloaded. This is the standard for cloud native application.

Reply:
@brittlewis12 I finally got your point, PR #5594 address it, thanks for pointing me out this.